[IN-6][Reviews] Add moderator management page

laurenbarker opened the pull request:
## TODO
- [x] get adding a new moderator working
- [x] get removing moderator working
- [x] get edit controls working
- [x] get adding a moderator by email working
- [x] fix row loading screen
- [x] move all text to translation file
- [x] update the error messages
- [x] add pagination controls
- [x] sort results alphabetically (backend)
- [x] fix linting errors
- [x] disable users in search results who are already moderators
- [x] disable everything for non-admin moderators
- [x] force refresh if a user is added or if pagination changes
- [x] add tests

## Purpose
Allow preprint providers to add and remove moderators using the moderation app.

## Depends on
* https://github.com/CenterForOpenScience/osf.io/pull/8186 (released)
* https://github.com/CenterForOpenScience/ember-osf/pull/370

## Changes
* Added "Moderators" tab
* Added "Moderators" link on dashboard
* Changed "Moderation" tab to "Submissions"
* Changed "Moderation" to "Submissions" on dashboard
* Made the tabs a little better shaped
* Added all of the moderator management functionality

![moderator-management](https://user-images.githubusercontent.com/7131985/40891470-2414c91c-6754-11e8-929f-ecc54f33f6c5.gif)

## Testing
Should look and act like [prototype](https://marvelapp.com/7hg387e/) with a few exceptions:
* When a new user is selected it keeps the input field
* Using header row instead of page title
* "Add new moderator" button is now just "Add"
* Moderators (as opposed to admin) cannot see the Permissions/Add/Remove functionality
* "Remove" instead of "Delete" moderators

Known oddities:
* You **can't add super users as moderators through the interface** (they have access though)
* We don't have a way of knowing the order that the users were added so alphabetical will have to do. This means that users added at the end might be in a different place if the page is reloaded.

## Ticket
https://openscience.atlassian.net/browse/IN-6

laurenbarker committed:
Fix adding, updating, removing requests

Fix skeleton screens

laurenbarker committed:
Add unregistered moderator form

laurenbarker committed:
Fix removing a moderator

Add loading indicator to buttons

laurenbarker committed:
Allow adding unregistered moderators

laurenbarker committed:
Add alt attribute to img

laurenbarker committed:
Resolve conflicts

laurenbarker committed:
Fix disabling row on edit

laurenbarker committed:
Disable search results for users who are already moderators

laurenbarker committed:
Disable add button when adding or editing

Moderation --> Submissions on dashboard

laurenbarker committed:
Use table headers

Remove everything but names for non-admin moderators

Add moderators to provider-links

laurenbarker committed:
Improve layout on small screens

laurenbarker committed:
Force reload if number of pages changes after add/delete

laurenbarker committed:
Add tests for moderator-row-component

Clean up css

aaxelb commented on app/components/moderator-list-add/component.js:
```diff
@@ -0,0 +1,136 @@
+            validator('presence', true),
+            validator('format', {
+                type: 'email',
+                regex: /^[-a-z0-9~!$%^&*_=+}{'?]+(\.[-a-z0-9~!$%^&*_=+}{'?]+)*@([a-z0-9_][-a-z0-9_]*(\.[-a-z0-9_]+)*\.(aero|arpa|biz|com|coop|edu|gov|info|int|mil|museum|name|net|org|pro|travel|mobi|[a-z][a-z])|([0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}\.[0-9]{1,3}))(:[0-9]{1,5})?$/i,
```
So IP hosts are fine, but you'll reject `anyone@dinosaurs.sexy`?

laurenbarker replied:
Oh lol I didn't write this I got it from https://github.com/CenterForOpenScience/ember-osf-preprints/blob/develop/app/components/unregistered-contributor-form.js#L27

aaxelb commented on app/components/moderator-list-add/component.js:
```diff
@@ -0,0 +1,136 @@
+
+        try {
+            // Hack to disable search results for users who are already moderators
+            // ember-data doesn't like when you manipulate the output
```
Couldn't you take the ember-data output and do the `map` you have below?
```js
const users = yield store.query('user', ...);
return users.map(user => {
  const userData = { ... };
  if (...) {
    userData.disabled = true;
  }
  return userData;
});
```
 Or is there something I'm missing?

laurenbarker replied:
So I thought the same thing but Ember did not agree. I don't think I completely removed the ember stuff when I was mapping though so I'll try that.

aaxelb commented on app/components/moderator-list-add/template.hbs:
```diff
@@ -0,0 +1,96 @@
+<div class="moderators-row-container p-h-md p-v-sm">
+    <div class="moderator-name">
+        {{#if fetchData.isRunning}}
```
`searchUsers`?

laurenbarker replied:
Hm this might be bad copy pasta...

aaxelb commented on tests/integration/components/moderator-list-row/component-test.js:
```diff
@@ -0,0 +1,93 @@
+    },
+});
+
+test('it renders moderator-list-row admin view', function(assert) {
```
Tests!?

aaxelb commented on app/components/moderator-list-add/component.js:
```diff
@@ -0,0 +1,136 @@
+    actions: {
+        roleChanged(role) {
+            this.set('role', role);
+            for (const value of this.get('roleOptions')) {
```
Equivalent, using `findBy`:
```js
this.set('roleLabel', this.get('roleOptions').findBy('role', role).label);
```
(but also maybe `roleLabel` should be a CP)

aaxelb commented on app/components/moderator-list-row/component.js:
```diff
@@ -0,0 +1,81 @@
+    role: computed('moderator.permissionGroup', function() {
+        for (const value of this.get('roleOptions')) {
+            if (value.role === this.get('moderator.permissionGroup')) {
+                this.set('roleLabel', value.label);
```
Computed properties with side effects are generally a bad idea, since ember will do all it can to avoid running it. Seems like it would make more sense for `roleLabel` to be a CP itself?

laurenbarker replied:
Yeah...the issue is I need role to be computed on the initial load and then set after that. I tried having it in `didRecieveAttrs` but that was triggered every time role changed. I think it would work in `init` though. Agree that it should not be a computed.

aaxelb replied:
So could it be a [oneWay](https://www.emberjs.com/api/ember/3.0/functions/@ember%2Fobject%2Fcomputed/oneWay) (or its more understandable alias, `reads`)?
```js
role: reads('moderator.permissionGroup')
```

aaxelb commented on app/components/moderator-list-row/component.js:
```diff
@@ -0,0 +1,81 @@
+    },
+
+    _setRole(role) {
+        this.set('role', role);
```
`role` is a computed property, it's confusing to also set it...

aaxelb commented on app/components/moderator-list-row/component.js:
```diff
@@ -0,0 +1,81 @@
+
+    _setRole(role) {
+        this.set('role', role);
+        for (const value of this.get('roleOptions')) {
```
Could be shorter with `findBy`, again.

aaxelb commented on app/controllers/preprints/provider/moderators.js:
```diff
@@ -0,0 +1,216 @@
+
+            yield moderatorInstance.destroyRecord({ adapterOptions: { provider: this.get('theme.provider.id') } });
+
+            const allModerators = yield this.get('store').peekAll('moderator');
```
`peekAll` is synchronous; the `yield` does nothing.

aaxelb commented on app/controllers/preprints/provider/moderators.js:
```diff
@@ -0,0 +1,216 @@
+        this.set('roleOptions', [
+            {
+                role: 'admin',
+                label: 'Admin',
```
i18n?

laurenbarker replied:
Oops, good catch!

aaxelb commented on app/styles/modules/_provider.scss:
```diff
@@ -1,3 +1,4 @@
 @import 'provider/settings';
 @import 'provider/setup';
+@import 'provider/moderators';
```
Highly recommend using ember-component-css [for route stylesheets too](https://github.com/ebryn/ember-component-css#usage-with-routes), rather than this `modules` setup.

Also, pod routes.

laurenbarker replied:
Little out of scope but will create a [ticket](https://openscience.atlassian.net/browse/IN-219) because I agree with you.

aaxelb commented on app/controllers/preprints/provider/moderators.js:
```diff
@@ -0,0 +1,216 @@
+    deleteModerator: task(function* (id) {
+        try {
+            // queryRecord doesn't play nice with the API
+            // findRecord with delete won't accept a payload
```
Could you pass `moderator` instead of `moderator.id` and do `moderator.destroyRecord()` right off the bat? Or is it important to double-check that it wasn't already deleted?

laurenbarker replied:
I'm not sure if I tried that -- fingers crossed 🤞

aaxelb reviewed (commented):
🍰

laurenbarker committed:
Move role options to translations file

Remove check for non-existent data loading

Remove unnecessary yield

Refactor role/roleLabel

Pass in moderatorInstance for removing and updating

Refactor searchUsers to use ember-data in a way that doesn't break

laurenbarker committed:
Remove unused var

laurenbarker committed:
Add tests for moderators controller

laurenbarker committed:
Add unit tests for moderator-list-row

laurenbarker committed:
Add unit tests for moderator-list-add component

laurenbarker committed:
Add integration tests for moderator-list-add

laurenbarker committed:
Add acceptance tests for moderators route

alexschiller commented on app/controllers/preprints/provider/moderators.js:
```diff
@@ -0,0 +1,200 @@
+            this.get('toast').error(this.get('i18n').t('moderators.deleteModeratorError'));
+            return false;
+        } finally {
+            this.set('editingModerator', false);
```
Edgy: 
<img width="1207" alt="screen shot 2018-04-20 at 11 37 54 am" src="https://user-images.githubusercontent.com/1322421/39062234-f28a1cd0-4494-11e8-98a1-559d3eef24aa.png">
Should the moderators be reloaded or `moderatorIds` be updated to prevent deleted mods from being disabled in search?

alexschiller commented on app/components/moderator-list-add/template.hbs:
```diff
@@ -0,0 +1,90 @@
+                selected=selectedUser
+                onchange=(action (mut selectedUser)) as |user|}}
+                <img class="m-xs" src="{{user.profileImage}}" height=25 width=25 alt="{{user.fullName}} gravatar">
+                <span>{{user.fullName}}</span>
```
Should this truncate? my full name (albeit all caps) goes onto a second line.
<img width="246" alt="screen shot 2018-04-20 at 12 06 58 pm" src="https://user-images.githubusercontent.com/1322421/39061853-aefa1cfa-4493-11e8-9e42-fb8a298e5606.png">

laurenbarker replied:
Can scroll within the dropdown if super long names:
![screen shot 2018-04-23 at 10 03 22 am](https://user-images.githubusercontent.com/7131985/39131545-adf2b4e2-46dd-11e8-9b58-0bc518911131.png)

alexschiller commented on app/components/moderator-list-row/template.hbs:
```diff
@@ -0,0 +1,75 @@
+    {{#if isAdmin}}
+        {{#bs-dropdown classNames="pull-right role-dropdown" as |dd|}}
+            {{#dd.button classNames="btn btn-default dropdown-button" disabled=disableRoleSelect}}
+                {{roleLabel}}
```
So it looks like mods only see the mod name list-- should they at least be able to see what role each person has (disabled button maybe)?  Seems they can see that data via the api, so it should be possible. If desirable, this can be a deferred improvement.

User story: I'm a moderator for a large preprint service and I want to be removed as a mod/upgraded to an admin but don't know who to contact.

laurenbarker replied:
It was decided in demo (and then not commented anywhere) that moderators would only see a list of names. It's definitely not a bad idea for moderators to see the the role though.

I'll follow up with product to make sure I'm remembering the decision correctly.

alexschiller replied:
:+1: Sounds good. It's a minor improvement, so nbd... that page just felt empty.

alexschiller commented on app/components/moderator-list-add/template.hbs:
```diff
@@ -0,0 +1,90 @@
+        {{/dd.menu}}
+    {{/bs-dropdown}}
+    <div class="row-controls">
+        <button class="btn btn-default" {{action 'cancel'}} disabled={{addModerator.isRunning}}>
```
Cancels get gnarly on small screen sizes... one of those, "probably not a priority, but..."
<img width="399" alt="screen shot 2018-04-20 at 12 21 18 pm" src="https://user-images.githubusercontent.com/1322421/39062538-d44bd0fa-4495-11e8-859e-ec74e9b9e469.png">
<img width="402" alt="screen shot 2018-04-20 at 11 38 31 am" src="https://user-images.githubusercontent.com/1322421/39062550-dc0f67ac-4495-11e8-893d-120bc5f49abe.png">

laurenbarker replied:
![screen shot 2018-04-23 at 10 01 24 am](https://user-images.githubusercontent.com/7131985/39131406-53acdfc6-46dd-11e8-9ecb-ac555aae27e4.png)

alexschiller commented on app/components/moderator-list-add/component.js:
```diff
@@ -0,0 +1,123 @@
+        validators: [
+            validator('presence', true),
+            validator('length', {
+                min: 3,
```
This should probably have a max as well, db max is set to 255

alexschiller commented on app/locales/en/translations.js:
```diff
@@ -230,6 +247,14 @@ export default {
             },
         },
+        moderatorListAdd: {
+            inviteText: 'invite by email',
```
stupid minor; when this text is highlighted it looks really cramped. Not sure if there is an easy fix. If you know one, do it, if not, leave it.
<img width="207" alt="screen shot 2018-04-20 at 11 33 11 am" src="https://user-images.githubusercontent.com/1322421/39062770-90c19b7a-4496-11e8-9b75-2c1705134832.png">

laurenbarker replied:
I can make it a little better at least:
![screen shot 2018-04-20 at 4 16 03 pm](https://user-images.githubusercontent.com/7131985/39071732-2eccefc6-44b6-11e8-8585-a492c60e66a4.png)

alexschiller replied:
Perfect ⭐️

alexschiller commented on app/components/moderator-list-add/component.js:
```diff
@@ -0,0 +1,123 @@
+                    profileImage: user.__data.links.profile_image,
+                    id: user.id,
+                };
+                if (this.get('moderatorIds').indexOf(user.id) > -1) {
```
I think this can use the es6 Array.includes
```javascript
if (this.get('moderatorIds').includes(user.id)) {
```

https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/includes

alexschiller reviewed (changes requested):
Couple things to look into.
At a minimum, add max length validation for full names. Open to thoughts/new tickets for other things as appropriate.

laurenbarker committed:
Use includes instead of indexOf

Add max length validation

laurenbarker committed:
Update wrapping for small screens

Add space between highlight and text

laurenbarker committed:
Update moderator ids after removing a moderator

alexschiller reviewed (changes requested):
As expected, conflicts ⚔️

laurenbarker committed:
Add another line to placeholder

Remove unused translations

laurenbarker committed:
Pin to most recent release/next-interfaces for ember-osf

alexschiller reviewed (approved):
🍰
